llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -3206,14 +3219,40 @@ NVPTXTargetLowering::LowerSTOREVector(SDValue Op, SelectionDAG &DAG) const {
     EVT EltVT = ValVT.getVectorElementType();
     unsigned NumElts = ValVT.getVectorNumElements();
 
+    if (UpsizeElementTypes) {
+      switch (ValVT.getSimpleVT().SimpleTy) {
+      default:
+        llvm_unreachable("Unexpected Vector Type");
+      case MVT::v8i8: // <2 x i8x4>
+        NumElts = 2;
+        EltVT = MVT::v4i8;
+        break;
+      case MVT::v8f16: // <4 x f16x2>
+        NumElts = 4;
+        EltVT = MVT::v2f16;
+        break;
+      case MVT::v8bf16: // <4 x bf16x2>
+        NumElts = 4;
+        EltVT = MVT::v2bf16;
+        break;
+      case MVT::v8i16: // <4 x i16x2>
+        NumElts = 4;
+        EltVT = MVT::v2i16;
+        break;
+      case MVT::v16i8: // <4 x i8x4>
+        NumElts = 4;
+        EltVT = MVT::v4i8;
+        break;
+      }
```

Comment:
This can be generalized to 
```
NPerWord = 32/EltVT.getSizeInBits(); // elements to pack in one word.
EltVT = getVectorVT(EltVT, NPerWord); // word-sized vector.
NumElts = NumElts/NPerWord; // number of word-sized vectors.
```

Reply:
Good suggestion. Would you still recommend hoisting this to a helper function now that we've reduced it to three lines?

Reply:
I guess. Keeping the magic in one place is a good thing. The helper could return `std::pair(NumElts, EltVT)` and structured binding lets us do it with virtually no boilerplate overhead.
```
auto helper = [](int N, EVT VT) {
  return std::pair(newN), newVT);
};
...
auto [N, VT] = helper(oldN, oldVT);
```

Reply:
Not super familiar with this syntax, do you know how to get around this issue? The problem is that we want to reassign the same variables that we pass into the lambda, rather than create new ones.
![image](https://github.com/user-attachments/assets/b97c7d57-d314-45d2-9962-6e98a93a9a8a)
![image](https://github.com/user-attachments/assets/0ac19cfc-bcc9-487d-bce1-3f51859a6420)

Reply:
With my most recent push, I believe I resolved everything except for this. I have a working solution with the lambda but it isn't very clean, if you have suggestions I would love to hear them.